File: paleo_workbench/workflow/interpretation/algorithm_registry.py
"""AlgorithmSpec registry — 算法能力唯一权威（V9 ADR-5）。

此前算法能力知识散落 ≥5 处（UI 标签表、engine id 映射、capabilities 矩阵、
对话框硬编码、harness 默认列表），且没有任何地方声明
schema/不确定度/CRS/约束支持。本模块收敛为单一注册表：

* ``algorithm_id`` = engine id（kriging/idw/constrained_idw/spline/
  directional/linear/nearest/rbf + factor_fusion）；
* ``aliases`` 覆盖 UI 中文标签与历史写法（canonical 解析）；
* ``supported_constraints`` 投影自 :mod:`constraint_capabilities`
  （engine 约束词汇，约束产品负责地质词汇→engine 词汇映射）；
* ``produces_uncertainty``/``requires_crs``/``requires_unit``/
  ``supports_cancel`` 为显式声明，UI/Agent/Workflow 不再手写判断。

职责边界（goal §32）：Algorithm Library = how to compute；
Template Library（mapping/geological_pipeline/templates）= how to present。
本注册表绝不描述制图模板。
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from paleo_workbench.workflow.constraint_capabilities import (
    ConstraintKind,
    Support,
    capabilities_for_method,
)


@dataclass(frozen=True)
class AlgorithmSpec:
    """一个科学算法的能力声明。"""

    algorithm_id: str
    family: str  # "interpolation" | "fusion"
    display_label: str
    aliases: tuple[str, ...] = ()
    #: UI 方法列表标签（历史值保持不变——tokens.INTERPOLATION_METHODS 派生源）。
    ui_label: str = ""
    #: 轻量参数 schema（JSON-schema 子集；供 UI/harness 校验，不承载科学）。
    parameter_schema: dict[str, Any] = field(default_factory=dict)
    #: 约束支持（engine 词汇；由 constraint_capabilities 权威投影）。
    supported_constraints: dict[str, tuple[str, str]] = field(default_factory=dict)
    supports_cancel: bool = True
    #: 插值/融合是否要求输入声明 CRS（未声明绝不猜测——诚实拒绝/降级）。
    requires_crs: bool = True
    requires_unit: bool = False
    produces_uncertainty: bool = False
    #: 后端（engine/库实现标识）。
    backend: str = ""
    prerequisites: tuple[str, ...] = ()
    notes: str = ""

    def constraint_support(self, kind: ConstraintKind) -> tuple[Support, str]:
        raw = self.supported_constraints.get(kind.value)
        if raw is None:
            return Support.UNSUPPORTED, "not consumed by this method"
        return Support(raw[0]), raw[1]

    def to_dict(self) -> dict[str, Any]:
        return {
            "algorithm_id": self.algorithm_id,
            "family": self.family,
            "display_label": self.display_label,
            "ui_label": self.ui_label or self.display_label,
            "aliases": list(self.aliases),
            "parameter_schema": dict(self.parameter_schema),
            "supported_constraints": {
                kind: list(entry) for kind, entry in self.supported_constraints.items()
            },
            "supports_cancel": self.supports_cancel,
            "requires_crs": self.requires_crs,
            "requires_unit": self.requires_unit,
            "produces_uncertainty": self.produces_uncertainty,
            "backend": self.backend,
            "prerequisites": list(self.prerequisites),
        }


def _from_capabilities(
    algorithm_id: str,
    label: str,
    *,
    aliases: tuple[str, ...] = (),
    ui_label: str = "",
    parameter_schema: dict[str, Any] | None = None,
    produces_uncertainty: bool = False,
    backend: str = "",
    notes: str = "",
) -> AlgorithmSpec:
    try:
        caps = capabilities_for_method(algorithm_id)
    except KeyError:
        caps = None
    support = {}
    if caps is not None:
        for kind, (level, note) in caps.support.items():
            support[kind.value] = (level.value, note)
    return AlgorithmSpec(
        algorithm_id=algorithm_id,
        family="interpolation",
        display_label=label,
        aliases=aliases,
        ui_label=ui_label or label,
        parameter_schema=parameter_schema or {},
        supported_constraints=support,
        produces_uncertainty=produces_uncertainty,
        backend=backend,
        prerequisites=tuple(caps.prerequisites) if caps else (),
        notes=notes,
    )


#: UI 方法列表顺序（tokens.INTERPOLATION_METHODS 的注册表权威来源）。
UI_INTERPOLATION_METHODS: tuple[str, ...] = (
    "kriging", "idw", "constrained_idw", "spline", "directional",
)

ALGORITHMS: dict[str, AlgorithmSpec] = {
    spec.algorithm_id: spec
    for spec in (
        _from_capabilities(
            "kriging", "克里金",
            ui_label="克里金",
            aliases=("克里金", "克里金(MVP·线性)", "ordinary_kriging", "ok"),
            parameter_schema={
                "type": "object",
                "properties": {
                    "grid_n": {"type": "integer", "minimum": 8, "maximum": 2000},
                    "variogram_model": {
                        "type": "string",
                        "enum": ["spherical", "exponential", "gaussian"],
                    },
                },
            },
            produces_uncertainty=True,  # kriging variance grid
            backend="kriging",
        ),
        _from_capabilities(
            "idw", "IDW 反距离加权",
            ui_label="IDW",
            aliases=("IDW", "反距离加权", "IDW反距离加权"),
            parameter_schema={
                "type": "object",
                "properties": {
                    "grid_n": {"type": "integer", "minimum": 8, "maximum": 2000},
                    "power": {"type": "number", "minimum": 0.5, "maximum": 8.0},
                },
            },
            backend="idw",
        ),
        _from_capabilities(
            "constrained_idw", "约束IDW",
            ui_label="约束IDW",
            aliases=("约束IDW", "约束反距离加权", "constrained idw"),
            parameter_schema={
                "type": "object",
                "properties": {
                    "grid_n": {"type": "integer", "minimum": 8, "maximum": 2000},
                    "power": {"type": "number", "minimum": 0.5, "maximum": 8.0},
                },
            },
            backend="constrained_idw",
        ),
        _from_capabilities(
            "spline", "样条 (CloughTocher)",
            ui_label="样条",
            aliases=("样条", "spline", "cubic", "样条插值"),
            backend="cubic",
        ),
        _from_capabilities(
            "directional", "方向趋势",
            ui_label="方向趋势",
            aliases=("方向趋势", "方向趋势面", "directional trend"),
            parameter_schema={
                "type": "object",
                "properties": {
                    "azimuth_deg": {"type": "number", "minimum": 0, "maximum": 360},
                    "semi_major": {"type": "number", "exclusiveMinimum": 0},
                    "semi_minor": {"type": "number", "exclusiveMinimum": 0},
                },
            },
            backend="directional",
        ),
        _from_capabilities("linear", "线性插值", aliases=("线性插值", "linear")),
        _from_capabilities("nearest", "最近邻", aliases=("最近邻", "nearest")),
        _from_capabilities("rbf", "RBF 多二次", aliases=("RBF", "rbf", "RBF 多二次")),
        AlgorithmSpec(
            algorithm_id="factor_fusion",
            family="fusion",
            display_label="多因素证据融合",
            aliases=("factor_fusion", "融合", "多因素融合", "weighted_evidence"),
            parameter_schema={
                "type": "object",
                "properties": {
                    "kind": {
                        "type": "string",
                        "enum": ["weighted_evidence", "rule_based"],
                    },
                    "class_thresholds": {"type": "array", "items": {"type": "number"}},
                },
            },
            # 融合的约束支持=空（约束在插值阶段消费；融合输入是 factor grids）。
            supported_constraints={},
            requires_crs=True,  # P0-5：混合声明纪律拒绝；全未声明 qc 降级
            produces_uncertainty=True,  # confidence grid + 方差传播（有方差输入时）
            backend="factor_fusion",
            notes="不确定性=coverage×agreement 置信格 + 公共支撑方差传播",
        ),
    )
}


def _rebuild_alias_index() -> None:
    _ALIAS_INDEX.clear()
    for spec in ALGORITHMS.values():
        _ALIAS_INDEX[spec.algorithm_id.lower()] = spec.algorithm_id
        for alias in spec.aliases:
            _ALIAS_INDEX[alias.lower()] = spec.algorithm_id


def register_algorithm(spec: AlgorithmSpec) -> None:
    """注册（或替换）算法声明（测试/扩展用）；别名索引同步重建。"""
    ALGORITHMS[spec.algorithm_id] = spec
    _rebuild_alias_index()


def get_algorithm(algorithm_id: str) -> AlgorithmSpec | None:
    return ALGORITHMS.get(str(algorithm_id or "").strip())


#: 别名 → algorithm_id（注册时经 _rebuild_alias_index 重建）。
_ALIAS_INDEX: dict[str, str] = {}
_rebuild_alias_index()


def canonical_algorithm_id(label_or_id: str) -> str:
    """任意 UI 标签/历史写法/engine id → 规范 algorithm_id。

    无法识别 → ValueError（不静默回退到默认方法——那正是词表漂移的来源）。
    """
    text = str(label_or_id or "").strip()
    if not text:
        raise ValueError("empty algorithm reference")
    resolved = _ALIAS_INDEX.get(text.lower())
    if resolved is None:
        raise ValueError(f"unknown algorithm: {label_or_id!r}")
    return resolved


def display_label(algorithm_id: str) -> str:
    spec = get_algorithm(canonical_algorithm_id(algorithm_id))
    return spec.display_label if spec else str(algorithm_id)


def ui_interpolation_methods() -> list[str]:
    """UI 方法列表（中文标签，按注册表顺序）——tokens 表的权威来源。"""
    return [
        (ALGORITHMS[a].ui_label or ALGORITHMS[a].display_label)
        for a in UI_INTERPOLATION_METHODS
        if a in ALGORITHMS
    ]


def interpolation_algorithm_labels() -> dict[str, str]:
    """UI 标签 → algorithm_id（供既有 label→engine 映射统一派生）。"""
    return {
        (ALGORITHMS[a].ui_label or ALGORITHMS[a].display_label): a
        for a in UI_INTERPOLATION_METHODS
        if a in ALGORITHMS
    }

File: libs/workflow_interpretation/src/algorithm_registry.cpp
// CONV-32 — C++ port of
// paleo_workbench/workflow/interpretation/algorithm_registry.py.
//
// The single authority for algorithm capability declarations: the frozen
// 9-entry ALGORITHMS table (insertion order kriging, idw, constrained_idw,
// spline, directional, linear, nearest, rbf, factor_fusion), the case-folded
// alias index, and the UI method-list derivations. Verbatim-string contract:
// labels/aliases/notes are byte-identical to the Python source; the oracle
// fixture (tools/oracle/generate_workflow_interpretation_registry_fixtures.py)
// freezes the real Python outputs this port must reproduce.
//
// Port notes (Python -> C++):
//  * ValueError parity: the frozen header declares canonical_algorithm_id
//    throws "AlgorithmValueError" but does not export the type, so it lives
//    here as a std::invalid_argument subclass with python_class()
//    == "ValueError". Callers catch std::invalid_argument; the message text
//    is the oracle ("empty algorithm reference" / "unknown algorithm:
//    '<repr of the ORIGINAL unstripped ref>'").
//  * ALGORITHMS insertion order cannot be preserved: the frozen header
//    exposes std::map<std::string, AlgorithmSpec> (key-sorted). No Python
//    observable depends on that order — the only iteration is
//    _rebuild_alias_index, whose per-spec key writes never collide across
//    algorithms, and the UI lists derive from UI_INTERPOLATION_METHODS, not
//    ALGORITHMS. Membership (order-insensitive) is the frozen contract.
//  * supported_constraints projects from capabilities_for_method in the
//    caps dict encounter order. MethodCapabilities::support is a
//    std::map<ConstraintKind, …> ordered by enum, which equals the Python
//    encounter order for every multi-cell method (constrained_idw:
//    boundary_mask < barrier < direction < anisotropy < trend;
//    directional: direction < anisotropy < trend). AlgorithmSpec stores the
//    projection keyed by kind VALUE string (frozen header type), so to_dict
//    emits those keys lexicographically — the Python dict order differs;
//    key order is not observable through the oracle's semantic compare.
//  * factor_fusion is a LITERAL spec: supported_constraints empty and
//    prerequisites empty (NOT projected from the capabilities matrix).
#include <pwb/workflow_interpretation/algorithm_registry.hpp>

#include <algorithm>
#include <stdexcept>
#include <utility>

namespace pwb::workflow_interpretation {
namespace {

// Python ValueError parity (see header comment). Defined here because the
// frozen header does not export the exception type; catch via its
// std::invalid_argument base.
class AlgorithmValueError : public std::invalid_argument {
public:
    explicit AlgorithmValueError(const std::string& what_arg)
        : std::invalid_argument(what_arg) {}
    [[nodiscard]] const char* python_class() const noexcept {
        return "ValueError";
    }
};

// Python str.strip() — ASCII whitespace only (no U+00A0/U+3000 forms exist
// in the frozen vocabulary).
std::string strip(const std::string& s) {
    const auto ws = [](unsigned char c) {
        return c == ' ' || c == '\t' || c == '\n' || c == '\r' || c == '\f' ||
               c == '\v';
    };
    std::size_t b = 0;
    std::size_t e = s.size();
    while (b < e && ws(static_cast<unsigned char>(s[b]))) ++b;
    while (e > b && ws(static_cast<unsigned char>(s[e - 1]))) --e;
    return s.substr(b, e - b);
}

// Python str.lower() for this vocabulary (ASCII letters only; CJK and the
// '·' separator are lower()-invariant).
std::string lower_ascii(const std::string& s) {
    std::string out = s;
    std::transform(out.begin(), out.end(), out.begin(), [](unsigned char c) {
        return static_cast<char>(c >= 'A' && c <= 'Z' ? c - 'A' + 'a' : c);
    });
    return out;
}

// Python repr() of a str for the frozen inputs (single quotes; backslash
// and single quote escaped). CPython would switch to double quotes when the
// payload contains ' but no "; no frozen input hits that case.
std::string py_repr_str(const std::string& s) {
    std::string out = "'";
    for (const char c : s) {
        if (c == '\\' || c == '\'') out += '\\';
        out += c;
    }
    out += '\'';
    return out;
}

Json schema(const std::string& text) { return Json::parse(text); }

struct RegistryState {
    std::map<std::string, AlgorithmSpec> algorithms;
    std::map<std::string, std::string> alias_index;
};

// _from_capabilities — capability projection (engine vocabulary).
AlgorithmSpec from_capabilities(const std::string& algorithm_id,
                                const std::string& label,
                                const std::vector<std::string>& aliases,
                                const std::string& ui_label,
                                const Json& parameter_schema,
                                bool produces_uncertainty,
                                const std::string& backend,
                                const std::string& notes) {
    std::optional<MethodCapabilities> caps;
    try {
        caps = capabilities_for_method(algorithm_id);
    } catch (const std::out_of_range&) {
        caps.reset();  // Python: except KeyError -> caps = None
    }
    AlgorithmSpec spec;
    spec.algorithm_id = algorithm_id;
    spec.family = "interpolation";
    spec.display_label = label;
    spec.aliases = aliases;
    spec.ui_label = ui_label.empty() ? label : ui_label;  // ui_label or label
    spec.parameter_schema = parameter_schema;
    if (caps.has_value()) {
        for (const auto& [kind, cell] : caps->support) {
            spec.supported_constraints[to_string(kind)] = {
                to_string(cell.level), cell.note};
        }
        spec.prerequisites = caps->prerequisites;
    }
    spec.produces_uncertainty = produces_uncertainty;
    spec.backend = backend;
    spec.notes = notes;
    return spec;
}

void rebuild_alias_index(RegistryState& state) {
    state.alias_index.clear();
    for (const auto& [id, spec] : state.algorithms) {
        state.alias_index[lower_ascii(id)] = id;
        for (const std::string& alias : spec.aliases) {
            state.alias_index[lower_ascii(alias)] = id;
        }
    }
}

RegistryState build_registry() {
    RegistryState state;
    // Insertion order kriging, idw, constrained_idw, spline, directional,
    // linear, nearest, rbf (factor_fusion appended below). std::map stores
    // key-sorted; only the rebuild iteration order is lost (collision-free —
    // see file header note).
    const std::vector<AlgorithmSpec> specs = {
        from_capabilities(
            "kriging", "克里金", {"克里金", "克里金(MVP·线性)", "ordinary_kriging", "ok"},
            "克里金",
            schema(R"({"type":"object","properties":{"grid_n":{"type":"integer","minimum":8,"maximum":2000},"variogram_model":{"type":"string","enum":["spherical","exponential","gaussian"]}}})"),
            true, "kriging", ""),
        from_capabilities(
            "idw", "IDW 反距离加权", {"IDW", "反距离加权", "IDW反距离加权"}, "IDW",
            schema(R"({"type":"object","properties":{"grid_n":{"type":"integer","minimum":8,"maximum":2000},"power":{"type":"number","minimum":0.5,"maximum":8.0}}})"),
            false, "idw", ""),
        from_capabilities(
            "constrained_idw", "约束IDW", {"约束IDW", "约束反距离加权", "constrained idw"},
            "约束IDW",
            schema(R"({"type":"object","properties":{"grid_n":{"type":"integer","minimum":8,"maximum":2000},"power":{"type":"number","minimum":0.5,"maximum":8.0}}})"),
            false, "constrained_idw", ""),
        from_capabilities(
            "spline", "样条 (CloughTocher)", {"样条", "spline", "cubic", "样条插值"},
            "样条", Json::object(), false, "cubic", ""),
        from_capabilities(
            "directional", "方向趋势", {"方向趋势", "方向趋势面", "directional trend"},
            "方向趋势",
            schema(R"({"type":"object","properties":{"azimuth_deg":{"type":"number","minimum":0,"maximum":360},"semi_major":{"type":"number","exclusiveMinimum":0},"semi_minor":{"type":"number","exclusiveMinimum":0}}})"),
            false, "directional", ""),
        from_capabilities("linear", "线性插值", {"线性插值", "linear"}, "",
                          Json::object(), false, "", ""),
        from_capabilities("nearest", "最近邻", {"最近邻", "nearest"}, "",
                          Json::object(), false, "", ""),
        from_capabilities("rbf", "RBF 多二次", {"RBF", "rbf", "RBF 多二次"}, "",
                          Json::object(), false, "", ""),
        // factor_fusion — literal spec (supported_constraints and
        // prerequisites are NOT projected from the capabilities matrix).
        [] {
            AlgorithmSpec spec;
            spec.algorithm_id = "factor_fusion";
            spec.family = "fusion";
            spec.display_label = "多因素证据融合";
            spec.aliases = {"factor_fusion", "融合", "多因素融合",
                            "weighted_evidence"};
            spec.ui_label = "";
            spec.parameter_schema = schema(
                R"({"type":"object","properties":{"kind":{"type":"string","enum":["weighted_evidence","rule_based"]},"class_thresholds":{"type":"array","items":{"type":"number"}}}})");
            spec.supported_constraints = {};
            spec.requires_crs = true;
            spec.produces_uncertainty = true;
            spec.backend = "factor_fusion";
            spec.notes = "不确定性=coverage×agreement 置信格 + 公共支撑方差传播";
            return spec;
        }(),
    };
    for (const AlgorithmSpec& spec : specs) {
        state.algorithms.emplace(spec.algorithm_id, spec);
    }
    rebuild_alias_index(state);
    return state;
}

RegistryState& state() {
    static RegistryState registry = build_registry();
    return registry;
}

}  // namespace

ConstraintSupport AlgorithmSpec::constraint_support(ConstraintKind kind) const {
    const auto it = supported_constraints.find(to_string(kind));
    if (it == supported_constraints.end()) {
        return {Support::UNSUPPORTED, "not consumed by this method"};
    }
    return {support_from(it->second.first), it->second.second};
}

Json AlgorithmSpec::to_dict() const {
    Json projected = Json::object();
    for (const auto& [kind, entry] : supported_constraints) {
        projected[kind] = Json::array({entry.first, entry.second});
    }
    Json out = Json::object();
    out["algorithm_id"] = algorithm_id;
    out["family"] = family;
    out["display_label"] = display_label;
    out["ui_label"] = ui_label.empty() ? display_label : ui_label;
    out["aliases"] = aliases;
    out["parameter_schema"] = parameter_schema;
    out["supported_constraints"] = projected;
    out["supports_cancel"] = supports_cancel;
    out["requires_crs"] = requires_crs;
    out["requires_unit"] = requires_unit;
    out["produces_uncertainty"] = produces_uncertainty;
    out["backend"] = backend;
    out["prerequisites"] = prerequisites;
    return out;
}

const std::map<std::string, AlgorithmSpec>& algorithms() {
    return state().algorithms;
}

void register_algorithm(const AlgorithmSpec& spec) {
    RegistryState& reg = state();
    reg.algorithms[spec.algorithm_id] = spec;  // replace-or-insert
    rebuild_alias_index(reg);
}

const AlgorithmSpec* get_algorithm(const std::string& id) {
    const auto it = state().algorithms.find(strip(id));
    return it == state().algorithms.end() ? nullptr : &it->second;
}

std::string canonical_algorithm_id(const std::string& ref) {
    const std::string text = strip(ref);
    if (text.empty()) {
        throw AlgorithmValueError("empty algorithm reference");
    }
    const auto it = state().alias_index.find(lower_ascii(text));
    if (it == state().alias_index.end()) {
        // repr of the ORIGINAL unstripped reference.
        throw AlgorithmValueError("unknown algorithm: " + py_repr_str(ref));
    }
    return it->second;
}

std::string display_label(const std::string& algorithm_id) {
    const AlgorithmSpec* spec = get_algorithm(canonical_algorithm_id(algorithm_id));
    return spec != nullptr ? spec->display_label : algorithm_id;
}

const std::vector<std::string>& ui_interpolation_method_ids() {
    // UI_INTERPOLATION_METHODS — Python insertion order.
    static const std::vector<std::string> ids = {"kriging", "idw",
                                                 "constrained_idw", "spline",
                                                 "directional"};
    return ids;
}

std::vector<std::string> ui_interpolation_methods() {
    std::vector<std::string> out;
    for (const std::string& id : ui_interpolation_method_ids()) {
        const auto it = state().algorithms.find(id);
        if (it == state().algorithms.end()) continue;
        const AlgorithmSpec& spec = it->second;
        out.push_back(spec.ui_label.empty() ? spec.display_label
                                            : spec.ui_label);
    }
    return out;
}

std::map<std::string, std::string> interpolation_algorithm_labels() {
    std::map<std::string, std::string> out;
    for (const std::string& id : ui_interpolation_method_ids()) {
        const auto it = state().algorithms.find(id);
        if (it == state().algorithms.end()) continue;
        const AlgorithmSpec& spec = it->second;
        out[spec.ui_label.empty() ? spec.display_label : spec.ui_label] = id;
    }
    return out;
}

}  // namespace pwb::workflow_interpretation
